invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -0,0 +1,341 @@
+                attn=attn,
+                hidden_states=hidden_states,
+                regional_ip_data=regional_ip_data,
+                query_length=query_length,
+                query=query,
+            )
+
+        # linear proj
+        hidden_states = attn.to_out[0](hidden_states)
+        # dropout
+        hidden_states = attn.to_out[1](hidden_states)
+
+        if input_ndim == 4:
+            hidden_states = hidden_states.transpose(-1, -2).reshape(batch_size, channel, height, width)
+
+        if attn.residual_connection:
+            hidden_states = hidden_states + residual
+
+        hidden_states = hidden_states / attn.rescale_output_factor
+        return hidden_states
+
+    def run_ip_adapters(
+        self,
+        attn: Attention,
+        hidden_states: torch.Tensor,
+        regional_ip_data: Optional[RegionalIPData],
+        query_length: int,  # TODO: just read from query?
+        query: torch.Tensor,
+    ) -> torch.Tensor:
+        if self._ip_adapter_attention_weights is None:
+            # If IP-Adapter is not enabled, then regional_ip_data should not be passed in.
+            assert regional_ip_data is None
+            return hidden_states
+
+        ip_masks = regional_ip_data.get_masks(query_seq_len=query_length)
+
+        assert (
+            len(regional_ip_data.image_prompt_embeds)
+            == len(self._ip_adapter_attention_weights)
+            == len(regional_ip_data.scales)
+            == ip_masks.shape[1]
+        )
+
+        for ipa_index, ip_hidden_states in enumerate(regional_ip_data.image_prompt_embeds):
+            # The batch dimensions should match.
+            # assert ip_hidden_states.shape[0] == encoder_hidden_states.shape[0]
+            # The token_len dimensions should match.
+            # assert ip_hidden_states.shape[-1] == encoder_hidden_states.shape[-1]
+
+            if self._ip_adapter_attention_weights[ipa_index].skip:
+                continue
+
+            ipa_weights = self._ip_adapter_attention_weights[ipa_index].ip_adapter_weights
+            ipa_scale = regional_ip_data.scales[ipa_index]
+            ip_mask = ip_masks[0, ipa_index, ...]
+
+            # Expected ip_hidden_state shape: (batch_size, num_ip_images, ip_seq_len, ip_image_embedding)
+            ip_key = ipa_weights.to_k_ip(ip_hidden_states)
+            ip_value = ipa_weights.to_v_ip(ip_hidden_states)
+
+            ip_hidden_states = self.run_attention(
+                attn=attn,
+                query=query,
+                key=ip_key,
+                value=ip_value,
+                attention_mask=None,
+            )
+
+            # Expected ip_hidden_states shape: (batch_size, query_seq_len, num_heads * head_dim)
+            hidden_states = hidden_states + ipa_scale * ip_hidden_states * ip_mask
+
+        return hidden_states
+
+    def prepare_attention_mask(
+        self,
+        attn: Attention,
+        attention_mask: Optional[torch.Tensor],
```

Comment:
`attention_mask` here should not be optional. It is what is being prepped. The function needs to be updated accordingly.